### ISSUE TITLE
Add a more structured demangling method

### ISSUE THREAD

**Issue by khuey** [EFanZh/ast-demangle#2]
`cpp_demangle` has a [`structured_demangle`](https://docs.rs/cpp_demangle/latest/cpp_demangle/struct.Symbol.html#method.structured_demangle) method that takes a [`limited form of a visitor trait`](https://docs.rs/cpp_demangle/latest/cpp_demangle/trait.DemangleWrite.html) that gets notified when certain AST nodes are traversed. The way the visitor trait is designed, emitted characters can be tied to certain AST nodes, allowing the consumer to do things like elide template arguments by default.

I'd like to introduce something similar to ast-demangle. Would you be satisfied with a limited visitor trait like `cpp_demangle` has or would you prefer something that's called for each AST node?

**Comment by EFanZh** [comment]
@khuey: Do you mean you need a method to traversal the symbol structure without building the `Symbol` object?

**Comment by khuey** [comment]
No, I want to produce the `Symbol` and then see which nodes output which characters (without duplicating that logic myself).

What I'm using right now is in https://github.com/Pernosco/ast-demangle if you want to take a look.
